### PR TITLE
Solving Project 2 Problems

### PR DESCRIPTION
**Pull request by paulghaddad**

### CODE REVIEW

File: src/project-2.js [hunk 1]
```diff
@@ -3,6 +3,7 @@
 const getBiggest = (x, y) => {
   // x and y are integers.  Return the larger integer
   // if they are the same return either one
+  return x > y ? x : y;
```
Comment (chichiwang) [inline src/project-2.js]: Opportunity for refactor - One-liner:
```javascript
const getBiggest = (x, y) => x > y ? x : y;
```

File: src/project-2.js [hunk 2]
```diff
@@ -11,15 +12,35 @@ const greeting = (language) => {
...
+    case 'Spanish':
+      return 'Hola!';
+    case 'Chinese':
+      return 'Ni Hao!';
+    default:
+      return 'Hello!';
+  }
 };
 
 const isTenOrFive = (num) => {
```
Comment (chichiwang) [inline src/project-2.js]: Opportunity for refactor - One-liner:
```javascript
const isTenOrFive = num => num === 10 || num === 5 ? true : false;
```

File: src/project-2.js [hunk 3]
```diff
@@ -44,62 +81,102 @@ const isPrime = (num) => {
...
+      return false;
+    }
+  }
+
+  return true;
 };
 
 const returnFirst = (arr) => {
   // return the first item from the array
+  return arr.shift();
```
Comment (chichiwang) [inline src/project-2.js]: Opportunity for refactor - mutation. `shift` mutates the parameter passed in and causes this method to have side effects. When working functionally, avoid all side-effects.

```javascript
const returnFirst = arr => arr[0];
```

File: src/project-2.js [hunk 4]
```diff
@@ -44,62 +81,102 @@ const isPrime = (num) => {
...
 };
 
 const returnLast = (arr) => {
   // return the last item of the array
+  return arr.pop();
```
Comment (chichiwang) [inline src/project-2.js]: Opportunity for refactor - mutation. `pop` mutates the parameter passed in and causes this method to have side effects. When working functionally, avoid all side-effects.

```javascript
const returnLast = arr => arr[arr.length - 1];
```

File: src/project-2.js [hunk 5]
```diff
@@ -44,62 +81,102 @@ const isPrime = (num) => {
...
 };
 
 const getArrayLength = (arr) => {
   // return the length of the array
+  return arr.length;
```
Comment (chichiwang) [inline src/project-2.js]: Opportunity for refactor - One-liner

```javascript
const getArrayLength = arr => arr.length;
```

File: src/project-2.js [hunk 6]
```diff
@@ -11,15 +12,35 @@ const greeting = (language) => {
...
+  if (num === 10 || num === 5) {
+    return true;
+  }
+
+  return false;
 };
 
 const isInRange = (num) => {
   // return true if num is less than 50 and greater than 20
+  if (num < 50 && num > 20) {
```
Comment (chichiwang) [inline src/project-2.js]: Opportunity for refactor - One-liner:

```javascript
const isInRange = num => num < 50 && num >20 ? true : false;
```

File: src/project-2.js [hunk 7]
```diff
@@ -44,62 +81,102 @@ const isPrime = (num) => {
...
 const getArrayLength = (arr) => {
   // return the length of the array
+  return arr.length;
 };
 
 const incrementByOne = (arr) => {
   // arr is an array of integers
   // increase each integer by one
   // return the array
+  const incrementedArray = arr.map(element => element += 1);
```
Comment (chichiwang) [inline src/project-2.js]: Opportunity for refactor - One-liner

```javascript
const incrementByOne = arr => arr.map(el => el + 1);
```

File: src/project-2.js [hunk 8]
```diff
@@ -44,62 +81,102 @@ const isPrime = (num) => {
...
+
+  return incrementedArray;
 };
 
 const addItemToArray = (arr, item) => {
   // add the item to the end of the array
   // return the array
+  arr.push(item);
```
Comment (chichiwang) [inline src/project-2.js]: Opportunity for refactor - mutation. `push` mutates the parameter passed in and causes this method to have side effects. When working functionally, avoid all side-effects.

```javascript
const addItemToArray = (arr, item) => [...arr, item];
```

File: src/project-2.js [hunk 9]
```diff
@@ -44,62 +81,102 @@ const isPrime = (num) => {
...
+
+  return arr;
 };
 
 const addItemToFront = (arr, item) => {
   // add the item to the front of the array
   // return the array
   // hint: use the array method .unshift
+  arr.unshift(item);
```
Comment (chichiwang) [inline src/project-2.js]: Opportunity for refactor - mutation. `unshift` mutates the parameter passed in and causes this method to have side effects. When working functionally, avoid all side-effects.

```javascript
const addItemToFront = (arr, item) => [item, ...arr];
```

File: src/project-2.js [hunk 10]
```diff
@@ -44,62 +81,102 @@ const isPrime = (num) => {
...
+  return arr;
 };
 
 const wordsToSentence = (words) => {
   // words is an array of strings
   // return a string that is all of the words concatenated together
   // spaces need to be between each word
   // example: ['Hello', 'world!'] -> 'Hello world!'
+
+  return words.join(' ');
```
Comment (chichiwang) [inline src/project-2.js]: Opportunity for refactor - One-liner

```javascript
const wordsToSentence = words => words.join(' ');
```

File: src/project-2.js [hunk 11]
```diff
@@ -44,62 +81,102 @@ const isPrime = (num) => {
...
 };
 
 const contains = (arr, item) => {
   // check to see if item is inside of arr
   // return true if it is, otherwise return false
+  const elementPresent = arr.indexOf(item);
```
Comment (chichiwang) [inline src/project-2.js]: Opportunity for refactor - One-liner

```javascript
const contains = (arr, item) => arr.indexOf(item) >= 0;
```

File: src/project-2.js [hunk 12]
```diff
@@ -44,62 +81,102 @@ const isPrime = (num) => {
...
+
+  return elementPresent >= 0;
 };
 
 const addNumbers = (numbers) => {
   // numbers is an array of integers.
   // add all of the integers and return the value
+  return numbers.reduce((sum, num) => sum + num);
```
Comment (chichiwang) [inline src/project-2.js]: Opportunity for refactor - One-liner

```javascript
const addNumbers = numbers => numbers.reduce((sum, num) => sum + num);
```

File: src/project-2.js [hunk 13]
```diff
@@ -44,62 +81,102 @@ const isPrime = (num) => {
...
 };
 
 const averageTestScore = (testScores) => {
   // testScores is an array.  Iterate over testScores and compute the average.
   // return the average
+  const sum = testScores.reduce((memo, num) => memo + num);
```
Comment (chichiwang) [inline src/project-2.js]: Opportunity for refactor - One-liner

```javascript
const averageTestScore = testScores => addNumbers(testScores)/testScores.length;
```

File: src/project-2.js [hunk 14]
```diff
@@ -44,62 +81,102 @@ const isPrime = (num) => {
...
+
+  return sum / testScores.length;
 };
 
 const largestNumber = (numbers) => {
   // numbers is an array of integers
   // return the largest integer
+  return numbers.reduce((largestItem, num) => {
```
Comment (chichiwang) [inline src/project-2.js:125]: Opportunity for refactor - One-liner

```javascript
const largestNumber = numbers.reduce((num1, num2) => Math.max(num1, num2));
```

**Comment by chichiwang** [comment]
@paulghaddad Very solid work - mostly just take advantage of lambda syntax wherever possible. It's hard to grok at first, but we may as well get used to functional notation.

On a side note, at our shop we've made the decision to skip arrow notation in favor of function notation except where we explicitly want the behavior of arrow notation.

We spent the period of a week sharing resources and a meeting discussing it: https://developer.ibm.com/node/2015/09/21/an-introduction-to-javascript-es6-arrow-functions/

Arrow notation applies a lexical `this` and binds the context in which it is declared. It is essentially `Function.proptotype.bind` behavior. We decided we should only use it when we explicitly want that behavior (React Class methods that are event handlers, for instance).

Arrow notations also do not benefit from hoisting. We've seen arguments in favor of and against this behavior.

Also anonymous arrow functions (and anonymous functions in general) leave a more confusing stacktrace.

Benchmarking arrow notation against the `function` keyword, I found that in browsers across the board, the `function` keyword outperforms arrow notation.

File: src/project-2.js (hunk 2 above)
Comment (paulghaddad) [reply src/project-2.js]: 👍  And even simpler: `const isTenOrFive = num => (num === 10 || num === 5);`

File: src/project-2.js (hunk 3 above)
Comment (paulghaddad) [reply src/project-2.js]: Great catch!

File: src/project-2.js (hunk 11 above)
Comment (paulghaddad) [reply src/project-2.js]: Used `includes` instead:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes

```
const contains = (arr, item) => arr.includes(item);
```

**Comment by paulghaddad** [comment]
@chichiwang Fantastic review! I learned a lot! Just updated the PR. Thank you so much! :)

File: src/project-2.js (hunk 11 above)
Comment (chichiwang) [reply src/project-2.js]: `includes` is fine for the exercise, but we will still be using `indexOf` for a long time in production (as it is not supported in our favorite browser as of yet).

File: src/project-2.js [hunk 15]
```diff
@@ -95,77 +81,43 @@ const isPrime = (num) => {
...
-  // add the item to the front of the array
-  // return the array
-  // hint: use the array method .unshift
-  arr.unshift(item);
+// return the length of the array
+const getArrayLength = arr => arr.length;
 
-  return arr;
-};
+const incrementByOne = arr => arr.map(element => element += 1);
```
Comment (chichiwang) [inline src/project-2.js]: No need to update `element` here, as the result is implicitly returned from the arrow function anyway. `element + 1` will work and skip the assignment.

**Pull request merged**